### PR TITLE
fix for status setting and ensuring we requeue on errors

### LINKED ISSUE #111: Bug: When AWS secret added after the zone the zone does not get into a ready state
**Steps**

Create a ManagedZone with no AWS secret present in the namespace
Create a secret afterwards
observe ManagedZone stays not ready with provider error

**Options**

- we should re-queue on this error

### PR DESCRIPTION
**Pull request by maleck13**
Fixes #111 

**Validation**

- create a managed zone with a provider secret ref that doesn't exist
- check status of zone
- create secret wait for reconcile to happen (depends on how long it has been failing as will back off)
- check status of zone

### CODE REVIEW

File: internal/controller/managedzone_controller.go [hunk 1]
```diff
@@ -106,40 +106,44 @@ func (r *ManagedZoneReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	// Publish the managed zone
 	err = r.publishManagedZone(ctx, managedZone)
 	if err != nil {
-		status = metav1.ConditionFalse
-		reason = "ProviderError"
-		message = fmt.Sprintf("The DNS provider failed to ensure the managed zone: %v", err)
-
-		err = r.Status().Update(ctx, managedZone)
```
Comment (maleck13) [inline internal/controller/managedzone_controller.go:113]: this wasn't doing anything before

File: internal/controller/managedzones.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import (
+	"context"
+	"fmt"
+
+	"github.com/kuadrant/dns-operator/api/v1alpha1"
+	"github.com/kuadrant/dns-operator/internal/provider"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func (r *ManagedZoneReconciler) publishManagedZone(ctx context.Context, managedZone *v1alpha1.ManagedZone) error {
```
Comment (maleck13) [inline internal/controller/managedzones.go:12]: I kinda want to put these on their own type so they have no access to the client

File: internal/controller/suite_test.go [hunk 3]
```diff
@@ -65,6 +64,10 @@ var dnsProviderFactory = &providerFake.Factory{
 	},
 }
 
+func SetDNSProvider(p provider.Provider) {
+	dnsProvider = p
+}
```
Comment (mikenairn) [inline internal/controller/suite_test.go:69]: I think we should avoid changing suite variables in tests. You won't be able to run tests in parallel if you do this as you will be modifying the provider that all specs will get back through the factory.

File: internal/controller/managedzone_controller_test.go [hunk 4]
```diff
@@ -100,5 +109,56 @@ var _ = Describe("ManagedZoneReconciler", func() {
...
+				if meta.IsStatusConditionTrue(mz.Status.Conditions, string(v1alpha1.ConditionTypeReady)) {
+					return fmt.Errorf("expected the ready condition to be false")
+				}
+				if mz.Status.NameServers != nil {
+					return fmt.Errorf("expected nameservers not to be set on provider error")
+				}
+				if mz.Status.RecordCount != 0 {
+					return fmt.Errorf("expected record count to be 0")
+				}
+				return nil
```
Comment (mikenairn) [inline internal/controller/managedzone_controller_test.go:145]: You might want to consider using gomega assertions here to be more consistent with all the other tests:

```.go
Eventually(func(g Gomega) {
	err = k8sClient.Get(ctx, client.ObjectKey{Namespace: mz.Namespace, Name: mz.Name}, mz)
	g.Expect(err).NotTo(HaveOccurred())
	g.Expect(mz.Status.Conditions).To(
		ContainElement(MatchFields(IgnoreExtras, Fields{
			"Type":               Equal(string(v1alpha1.ConditionTypeReady)),
			"Status":             Equal(metav1.ConditionTrue),
			"Reason":             Equal("Whatever this is meant to be"),
			"Message":            Equal("Whatever this is meant to be"),
			"ObservedGeneration": Equal(mz.Generation),
		})),
	)
	g.Expect(mz.Finalizers).To(ContainElement(ManagedZoneFinalizer))
	g.Expect(mz.Status.NameServers).To(BeNil())
	g.Expect(mz.Status.RecordCount).To(BeZero())
}, TestTimeoutMedium, time.Second).Should(Succeed())
```

File: internal/controller/suite_test.go (hunk 3 above)
Comment (maleck13) [reply internal/controller/suite_test.go:69]: Yeah just not really sure what other option we have

File: internal/controller/managedzone_controller.go [hunk 5]
```diff
@@ -186,21 +204,21 @@ func (r *ManagedZoneReconciler) deleteManagedZone(ctx context.Context, managedZo
...
 	if err != nil {
 		if strings.Contains(err.Error(), "was not found") || strings.Contains(err.Error(), "notFound") {
 			log.Log.Info("ManagedZone was not found, continuing", "managedZone", managedZone.Name)
 			return nil
 		}
-		return err
+		return ZoneErr{
+			code:    ZoneErrProvider,
+			message: fmt.Sprintf("Failed to delete from provider. Provider Error: %v", provider.SanitizeError(err)),
+		}
```
Comment (mikenairn) [inline internal/controller/managedzone_controller.go:221]: Is this preferable to just wrapping an error with a known type?

```.go
var (
   ErrProvider = errors.New("ProviderError")
)
...
if err != nil {
   return fmt.Errorf("%w, failed to get provider for the zone: %v", ErrProvider, provider.SanitizeError(err))
}
```

and using errors.Is where you want to check the type:

```go
if errors.Is(ErrProvider, err) {
   reason = ErrProvider.Error()
   message = err.Error()
}
```

File: internal/controller/suite_test.go (hunk 3 above)
Comment (maleck13) [reply internal/controller/suite_test.go:69]: We need a way to simulate errors from the provider IMO. Perhaps we create a follow up to add some form of error injection into the inmem provider?
Like if a test could set a response keyed to itself that would work

File: internal/controller/managedzone_controller_test.go [hunk 6]
```diff
@@ -100,5 +110,59 @@ var _ = Describe("ManagedZoneReconciler", func() {
 			Expect(err).To(HaveOccurred())
 			Expect(err.Error()).To(ContainSubstring("spec.domainName in body should match"))
 		})
+
+		It("status should move from false to true when provider error resolved", func() {
```
Comment (maleck13) [inline internal/controller/managedzone_controller_test.go:114]: @mikenairn I am not sure how great these tests are. If we would prefer I can remove them.

**Comment by maleck13** [comment]
@mikenairn if no objections will I go ahead and merge?

File: internal/controller/suite_test.go [hunk 7]
```diff
@@ -59,6 +59,10 @@ var testEnv *envtest.Environment
 var ctx context.Context
 var cancel context.CancelFunc
 
+func SetDNSProvider(p provider.Provider) {
+	dnsProvider = p
+}
+
```
Comment (mikenairn) [inline internal/controller/suite_test.go:65]: Think this can be removed

**Review by mikenairn — APPROVED** [review]

**Review by mikenairn — APPROVED** [review]

**Pull request merged**